### PR TITLE
[NCL-5483] Add websocket support to rest-client

### PR DESCRIPTION
**Pull request by michalovjan**
### Checklist:

* [ ] Have you added a note in the [CHANGELOG wiki](https://github.com/project-ncl/pnc/wiki/Changelog) for your change if user-facing?
* [ ] Have you added unit tests for your change?

### CODE REVIEW

File: rest-client/src/main/java/org/jboss/pnc/restclient/websocket/VertxWebSocketClient.java [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+     * number of attempted retries before client throws an exception
+     */
+    private int maximumRetries = 5;
+
+    private int numberOfRetries = 0;
+
+    /**
+     * a multiplier that increases delay between reconnect attempts
+     */
+    private float delayMultiplier = 1.5F;
```
Comment (thescouser89) [inline rest-client/src/main/java/org/jboss/pnc/restclient/websocket/VertxWebSocketClient.java]: I'd say just use as delayMultiplier the number '2'.

And with that you can optimize by doing a bitwise shift left to "multiply" by 2.

**Comment by thescouser89** [comment]
LGTM with my 1 tiny comment. With the CompletableFuture, I can chain callbacks after the results are obtained.

File: rest-client/src/main/java/org/jboss/pnc/restclient/websocket/VertxWebSocketClient.java (hunk 1 above)
Comment (michalovjan) [reply rest-client/src/main/java/org/jboss/pnc/restclient/websocket/VertxWebSocketClient.java]: For sure can change it to 2. With bitwise shift it forces the number to be 2 and doesn't allow for the number to change (currently can be set through constructor). Do we want to hard code it?

**Comment by michalovjan** [comment]
> LGTM with my 1 tiny comment. With the CompletableFuture, I can chain callbacks after the results are obtained.
> 
> Could you add a method for getting the live logs of a build? .

Oh yeah! Live logs! Completely forgot about them :D. 

> Same for creation of build configuration, and scm repository

Do I understand this correctly? To send every processed message/notification also to stdout? :)

**Comment by matejonnet** [comment]
Nice!

File: rest-client/src/main/java/org/jboss/pnc/restclient/websocket/VertxWebSocketClient.java (hunk 1 above)
Comment (janinko) [reply rest-client/src/main/java/org/jboss/pnc/restclient/websocket/VertxWebSocketClient.java]: I once heard some thing about premature optimization. Doesn't it apply here?

Comment (alexcreasy) [reply rest-client/src/main/java/org/jboss/pnc/restclient/websocket/VertxWebSocketClient.java]: errm, changing from 1.5 to 2 is a pretty big increase in backoff factor... isn't 1.3 usually the standard?

Comment (alexcreasy) [reply rest-client/src/main/java/org/jboss/pnc/restclient/websocket/VertxWebSocketClient.java]: actually I don't know where I got 1.3 from, my brain offered it up, but I can't find anything to back it up...

**Comment by thescouser89** [comment]
+1 to merge it! :)

File: rest-client/src/main/java/org/jboss/pnc/restclient/websocket/VertxWebSocketClient.java (hunk 1 above)
Comment (michalovjan) [reply rest-client/src/main/java/org/jboss/pnc/restclient/websocket/VertxWebSocketClient.java]: I've made that number up, so I am free to change it to whatever you want. We can also make it customizable through CLI config or smthing like that.

**Pull request merged**